### PR TITLE
Add snap support.

### PR DESCRIPTION
**Pull request by dingdang66686**

### CODE REVIEW

**Comment by mjakeman** [comment]
Hi, thanks for working on this!

At a high level it all looks good to me. I'll have to test this in some detail but I'm happy to keep the snap build files in-tree as an 'official' port.

One comment: could you please rename the app from `gnome-extension-manager` to simply `extension-manager`? We're not a GNOME core app so using according to the [App Naming Guidelines](https://developer.gnome.org/hig/guidelines/app-naming.html) we should avoid using 'GNOME' in the name.

**Review by mjakeman — CHANGES_REQUESTED** [review]

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+name: gnome-extension-manager # you probably want to 'snapcraft register <name>'
```
[the remark this answers is not in the document]
Comment (Tamaraa01x) [reply snap/snapcraft.yaml:1]: - [ ]

[the remark this answers is not in the document]
Comment (dingdang66686) [reply snap/snapcraft.yaml]: I have changed that.

**Review by Tamaraa01x — APPROVED** [review]

**Comment by dingdang66686** [comment]
This configuration is now usable.

**Comment by mjakeman** [comment]
Hi @dingdang66686, sorry for the delay in testing.

It seems snapcraft is [completely broken on the arm64 architecture](https://bugs.launchpad.net/snapcraft/+bug/1999321), at least on the two distributions I've tried it on, so I'm unable to test the snap at the moment. I'll try dig up an x86 laptop to test it with soon.

In regards to snap support, I notice you've got a package here: https://snapcraft.io/gnome-extension-manager. Did you want to continue maintaining this unofficially, or transfer ownership and have it maintained upstream (as part of this PR)?

**Comment by dingdang66686** [comment]
I have to mention that this application doesn't work in sandboxed mode. When apparmor is enabled, the dbus message is blocked. I have checked the official documentation but without luck.

**Comment by dingdang66686** [comment]
https://1drv.ms/u/s!AjQ7y8jf_bBFmIRY_zOSIdzXAMD-6w?e=CfKWrG @mjakeman Test this snap on your arm64 device.

**Comment by mjakeman** [comment]
Hi @dingdang66686,

I've finally had the chance to review and test this. I've fixed the issues I was having with arm snaps by using an Ubuntu VM image. On the whole, the package looks good.

There are a few points: 
 - Could you add 'arm64' to the supported architectures
 - The snap is currently in devmode. There's nothing preventing extension-manager from working within a sandbox; I believe it should be possible to use the [`dbus`](https://forum.snapcraft.io/t/the-dbus-interface/2038) interface to make it confined again?

**Review by mjakeman — CHANGES_REQUESTED** [review]
See above

**Comment by mjakeman** [comment]
Unfortunately, AppArmor prevents the app from using dbus, meaning that it can't install anything now.

image

There isn't anything stopping the app from working with confinement (by design), so I'd ask that the snap file declares the proper permissions so we can use strict confinement. Maybe this will help: https://forum.snapcraft.io/t/the-dbus-interface/2038

**Comment by dingdang66686** [comment]
As the document says, the dbus interface is only for communications between snaps. I don't think this will help.

**Comment by mjakeman** [comment]
> As the document says, the dbus interface is only for communications between snaps. I don't think this will help.

Yep, you're absolutely right. I misread that, apologies.

I've spent a lot of time looking at snap interfaces and how we can work around this for confinement purposes. Unfortunately, if there is a solution it does not look trivial (which is a bit odd, given that DBus is the primary means of achieving confinement under flatpak).

Thanks for your patience with this PR - I'm happy to take this as-is in time for the 0.4.1 release, and we can look at tightening things up later on.

**Review by mjakeman — APPROVED** [review]

**Comment by mjakeman** [comment]
Hi @dingdang66686, I'm looking at making a new release of Extension Manager in the next week or so. I'm happy to take ownership of the snap package and handle releases if you're agreeable?

**Comment by dingdang66686** [comment]
I'm glad to do so. 😁

**Comment by dingdang66686** [comment]
Maybe we should ask snap developers to add a new slot/plug for gnome extension managing.

**Comment by mjakeman** [comment]
Great! I've opened a request: https://forum.snapcraft.io/t/transfer-ownership-of-extension-manager-and-rename-package/33870?u=mjakeman

> Maybe we should ask snap developers to add a new slot/plug for gnome extension managing.

I think that's a good idea. Let's publish v0.4.1 using classic confinement so we have a working version out in the wild. We can ask for either a session-dbus interface, or maybe some kind of gnome-specific plug that whitelists `org.gnome.Shell.Extensions`?

**Comment by dingdang66686** [comment]
You can register a different snap named extension-manager. Then I will close my snap.

**Pull request merged**